File: Program's_Contributed_By_Contributors/Java_Programs/ValidParentheses.java
import java.util.Map;
import java.util.HashMap;
import java.util.Stack;

class ValidParentheses {
  public static boolean isValid(String str) {    
    Map<Character, Character> parenthesesMapping = new HashMap<>();
    parenthesesMapping.put('(', ')');
    parenthesesMapping.put('{', '}');
    parenthesesMapping.put('[', ']');

    Stack<Character> parentheses = new Stack<>();
    for(int i = 0; i < str.length(); i++) {
        char c = str.charAt(i);
        if(parenthesesMapping.containsKey(c)) {
            parentheses.push(c);    
        } else {
            if(parentheses.isEmpty()) {
                return false;
            }
            char target = parentheses.pop();
            if(!parenthesesMapping.containsKey(target) || parenthesesMapping.get(target) != c) {
                return false;
            }
        }
    }
    if(!parentheses.isEmpty()) {
        return false;
    }
    return true;
  }

  public static void main(String[] args) {
	Scanner sc = new Scanner(System.in);
	String str = sc.next();
	System.out.println(isValid(str));
    //System.out.println(isValid("([)]"));
  }
}

File: Program's_Contributed_By_Contributors/C++ Programs/Stacks/Valid_Parentheses.cpp
class Solution {
public:
    bool isValid(string s) {

        stack<char> st;
        int n = s.length();

        for(int i=0;i<n;i++)
        {
            if((s[i]==')' || s[i]=='}' ||s[i]==']') && (st.empty()))
            {
                return false;
            }

            if((s[i]==')' && st.top()=='(') || (s[i]=='}' && st.top()=='{') || (s[i]==']' && st.top()=='['))
            {
                st.pop();
            }
            else
                st.push(s[i]);
        }

        if(st.empty())
            return true;

        return false;
    }
};
